src/codegen.cpp

```diff
@@ -3843,6 +3844,40 @@ static jl_cgval_t emit_expr(jl_codectx_t &ctx, jl_value_t *expr)
     else if (head == boundscheck_sym) {
         return mark_julia_const(bounds_check_enabled(ctx, jl_true) ? jl_true : jl_false);
     }
+    else if (head == gc_preserve_sym) {
+        size_t nargs = jl_array_len(ex->args);
+        jl_cgval_t *argv = (jl_cgval_t*)alloca(sizeof(jl_cgval_t) * nargs);
+        for (size_t i = 0; i < nargs; ++i) {
+            argv[i] = emit_expr(ctx, args[i]);
+        }
+        size_t nargsboxed = 0;
+        Value **vals = (Value**)alloca(sizeof(Value *) * nargs);
+        for (size_t i = 0; i < nargs; ++i) {
+            if (!argv[i].isboxed) {
+                jl_error("Expr(:gc_preserve) only works on boxed values");
```

Comment:
We allow this for `ccall` so we should also just ignore it here?

Reply:
No, I don't think so. This doesn't make sense on non-boxed values and people may be surprised, by e.g.:
```
n = 100000
@gc_preserve n do_something(pointer_from_objref(n))
```
otherwise. With this framework, we can also add a version of pointer_form_objref that gives you the pointer as well as a lifetime token, but @vtjnash seemed to think we should get rid of pointer_form_objref for non-obviously-boxed values anyway.

Reply:
The ccall version also doesn't guarantee boxing of isbits objects if it is passed in as a gc root, i.e. the following def is invalid.

```julia
cconvert(::Type{Ref{Int}}, v::Int) = v
unsafe_convert(::Type{Ref{Int}}, v::Int) = Ptr{Int}(pointer_from_objref(v))
```

I think this is not so different from your example.

This mainly makes it very hard to write generic unsafe code in a safe way. The most obvious usecase that I'd like to support is to implement an unsafe C function in julia and being able to replace

```
ccall(:c_vertion, Void, (Ptr{Void},), obj)
```

with

```julia
root = cconvert(Ptr{Void}, obj)
@gc_preserve root begin
    ptr = unsafe_convert(Ptr{Void}, root)
    julia_vertion(ptr)
end
```

And not having to worry about the type of `root`.

This also makes optimization in inference harder since we might want to split the root just like what we do for ccall.

Also, in any case, this should emit an error instead of throwing one from codegen.

Reply:
> With this framework, we can also add a version of pointer_form_objref that gives you the pointer as well as a lifetime token

You mean `((a = cconvert(Ref{Any}, v)), unsafe_convert(Ref{Any}, a))`?

> get rid of pointer_form_objref for non-obviously-boxed values anyway.

Having it raise an error is certainly doable. As lone as we still have a more hidden version for debugging like `ccall(:jl_value_ptr, Ptr{Void}, (Any,), v)`

Reply:
> This mainly makes it very hard to write generic unsafe code in a safe way 

I think Ref already should already provide a way to handle this case:
```
julia> Base.unsafe_convert(Ref{Any}, Base.cconvert(Ref{Any}, 1.0))
Ptr{Any} @0x0000000117b5c1a0

julia> unsafe_load(ans)
1.0
```

Reply:
> I think Ref already should already provide a way to handle this case:

Not when you don't know the type of the input and not when `pointer_to_objref` isn't what you want to get (and it's not what my code above is doing anyway).

Reply:
If we disallow pointer_from_objref in the problematic case, I'm fine with ignoring this error here. If not, I think I'd be worried that people expect that
`@gc_preserve x use(pointer_from_objref(x))` will keep the pointer returned from `pointer_from_objref` valid.

Reply:
What's the difference between this and the `unsafe_convert` code I give? Both of them are writing code for unsafe code so the user should be already warned of possible issue and should read the doc carefully about the semantics.

Reply:
Ok, I'll remove the error.